File: scripts/age_grouping_examples.R
# scripts/age_grouping_examples.R
# Usage: edit DATA_FILE to point to your CSV; then run with Rscript scripts/age_grouping_examples.R
# Expects CSV with rows = samples and columns: proteins named starting with 'p', and columns: age, sex, batch, CRP (optional), SampleID (optional)

library(limma)
library(pheatmap)
# Optional: library(sva)

set.seed(12345)

# === User params ===
DATA_FILE <- "data/ihm_real_age_analysis.csv"  # real IHM protein matrix with derived age/sex/batch/CRP metadata
OUT_DIR <- "outputs"
PROT_PREFIX <- "p"
TOPN <- 50
# ====================

dir.create(OUT_DIR, showWarnings = FALSE, recursive = TRUE)

cat("Loading data:", DATA_FILE, "\n")
df <- read.csv(DATA_FILE, stringsAsFactors = FALSE)

# Some raw matrices append a trailing Group row for annotation; remove it defensively.
if (nrow(df) > 0 && any(grepl('^Group', trimws(as.character(df[[1]])), ignore.case = TRUE))) {
  df <- df[!grepl('^Group', trimws(as.character(df[[1]])), ignore.case = TRUE), , drop = FALSE]
}

# Age grouping: quartiles and custom
# quartiles
q_breaks <- quantile(df$age, probs = seq(0,1,0.25), na.rm = TRUE)
df$age_q <- cut(df$age, breaks = q_breaks, include.lowest = TRUE, labels = c("Q1","Q2","Q3","Q4"))
# custom bins example (edit as needed)
mybreaks <- c(-Inf, 50, 65, 80, Inf)
df$age_grp <- cut(df$age, breaks = mybreaks, include.lowest = TRUE, labels = c("<50","50-64","65-79","80+"))
# ordinal coding for trend
if(is.factor(df$age_q)) df$age_ord <- as.numeric(df$age_q) else df$age_ord <- as.numeric(factor(df$age_q))

# proteins
prot_cols <- grep(paste0('^', PROT_PREFIX), names(df), value = TRUE)
if(length(prot_cols) == 0) stop('No protein columns found with prefix ', PROT_PREFIX)

# expression matrix (rows=proteins, cols=samples)
expr_mat <- as.matrix(t(df[, prot_cols, drop = FALSE]))
colnames(expr_mat) <- if(!is.null(df$SampleID)) df$SampleID else paste0('S', seq_len(ncol(expr_mat)))

# design: use quartile factor (reference Q1) and covariates
# ensure factor ref
df$age_q <- relevel(factor(df$age_q), ref = "Q1")
design <- model.matrix(~ df$age_q + df$sex + df$batch + df$CRP)
colnames(design) <- make.names(colnames(design))

# limma fit
fit <- lmFit(expr_mat, design)
fit <- eBayes(fit)

# Find coef name for Q4 vs Q1 (example) or trend (age_ord)
# look for common age_q Q4 column name patterns
coef_name_q4 <- grep('age_qQ4', colnames(design), value = TRUE)
if(length(coef_name_q4) == 0) coef_name_q4 <- grep('Q4', colnames(design), value = TRUE)
if(length(coef_name_q4) == 0) coef_name_q4 <- NULL

# prefer trend test if present
coef_name_trend <- NULL
if('age_ord' %in% colnames(df)){
  design_trend <- model.matrix(~ df$age_ord + df$sex + df$batch + df$CRP)
  fit_trend <- lmFit(expr_mat, design_trend)
  fit_trend <- eBayes(fit_trend)
  # find numeric age_ord coefficient name in fit_trend
  coef_name_trend <- grep('age_ord', colnames(design_trend), value = TRUE)
  if(length(coef_name_trend)==0) coef_name_trend <- NULL
}

# If trend is available prefer it
if(!is.null(coef_name_trend)){
  res <- topTable(fit_trend, coef = coef_name_trend[1], number = Inf, adjust.method = 'BH')
  res$protein <- rownames(res)
  write.csv(res, file = file.path(OUT_DIR, 'limma_results_trend.csv'), row.names = FALSE)
  chosen <- head(rownames(res[order(res$adj.P.Val),]), n = TOPN)
} else {
  # attempt to find an age-related coefficient in the original design
  age_coefs <- grep('age_q|age', colnames(design), value = TRUE)
  age_coefs <- setdiff(age_coefs, '(Intercept)')
  if(length(age_coefs) > 0){
    res <- topTable(fit, coef = age_coefs[1], number = Inf, adjust.method = 'BH')
    res$protein <- rownames(res)
    write.csv(res, file = file.path(OUT_DIR, 'limma_results_age_group.csv'), row.names = FALSE)
    chosen <- head(rownames(res[order(res$adj.P.Val),]), n = TOPN)
  } else {
    stop('No age coefficient found in design matrix')
  }
}

# subset and z-score per protein
sub_expr <- expr_mat[chosen, , drop = FALSE]
scaled <- t(scale(t(sub_expr)))

ann_col <- data.frame(AgeQuartile = df$age_q, AgeBin = df$age_grp, Sex = df$sex, Batch = df$batch, CRP = df$CRP)
rownames(ann_col) <- colnames(scaled)

# heatmap
pheatmap(scaled, annotation_col = ann_col, show_rownames = TRUE, show_colnames = FALSE, clustering_method = 'complete', filename = file.path(OUT_DIR, 'age_diff_heatmap_limma.png'), width = 10, height = 8)

cat('R script finished. Results in', OUT_DIR, '\n')

File: scripts/age_grouping_examples.py
# scripts/age_grouping_examples.py
# Usage: edit DATA_FILE to point to your CSV; then run with python scripts/age_grouping_examples.py
# Expects CSV with rows = samples and columns: proteins named starting with 'p', and columns: age, sex, batch, CRP (optional), SampleID (optional)

import os
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from statsmodels.api import OLS
from statsmodels.tools.tools import add_constant
from statsmodels.stats.multitest import multipletests

SEED = 12345
np.random.seed(SEED)

# === User params ===
DATA_FILE = 'data/expr_sample_table.csv'  # change to your SomaScan/Olink CSV path
OUT_DIR = 'outputs'
PROT_PREFIX = 'p'   # set to '' to auto-detect protein columns (recommended for real SomaScan/Olink)
PROTEIN_LIST_FILE = 'data/top50_proteins.txt'  # optional, one protein per line; if present, used directly
TOP_N = 50  # number of top proteins to select for heatmap
REF_LEVEL = 'Q1'  # reference for quartiles
# ====================

os.makedirs(OUT_DIR, exist_ok=True)

print('Loading data:', DATA_FILE)
df = pd.read_csv(DATA_FILE)

# quartiles and custom bins
q_breaks = df['age'].quantile([0,0.25,0.5,0.75,1.0]).values
# avoid duplicate breaks (constant age) by jittering tiny amount if needed
if len(np.unique(q_breaks)) < len(q_breaks):
    q_breaks = np.linspace(df['age'].min()-1e-6, df['age'].max()+1e-6, 5)

df['age_q'] = pd.qcut(df['age'], q=4, labels=['Q1','Q2','Q3','Q4'])
# custom example
bins = [-np.inf, 50, 65, 80, np.inf]
labels = ['<50','50-64','65-79','80+']
df['age_grp'] = pd.cut(df['age'], bins=bins, labels=labels, include_lowest=True)
# ordinal coding
if pd.api.types.is_categorical_dtype(df['age_q']):
    df['age_ord'] = df['age_q'].cat.codes + 1
else:
    df['age_ord'] = pd.Categorical(df['age_q']).codes + 1

# protein columns
if PROT_PREFIX == '':
    # auto-detect protein columns: numeric columns excluding known metadata
    metadata_cols = set(['age','sex','batch','CRP','SampleID','age_q','age_grp','age_ord'])
    prot_cols = [c for c in df.columns if c not in metadata_cols and pd.api.types.is_numeric_dtype(df[c])]
else:
    prot_cols = [c for c in df.columns if c.startswith(PROT_PREFIX)]
if len(prot_cols) == 0:
    raise SystemExit('No protein columns found; set PROT_PREFIX or provide a protein list file')

# covariates
covar_cols = []
for c in ['sex','batch','CRP']:
    if c in df.columns:
        covar_cols.append(c)

# prepare base design matrix (one-hot encode categorical covariates drop_first)
X_base = pd.DataFrame(index=df.index)
for c in covar_cols:
    # treat non-numeric as categorical
    if not pd.api.types.is_numeric_dtype(df[c]):
        dummies = pd.get_dummies(df[c].astype(str), prefix=c, drop_first=True)
        X_base = pd.concat([X_base, dummies], axis=1)
    else:
        # coerce numeric, allow NaN
        X_base[c] = pd.to_numeric(df[c], errors='coerce')
# add intercept
X_base = add_constant(X_base, has_constant='add')

# age dummies for quartiles (drop_first -> Q1 reference)
age_dummies = pd.get_dummies(df['age_q'].astype(str), prefix='age_q', drop_first=True)
if age_dummies.shape[1] == 0:
    raise SystemExit('age_q dummy not created; check age data')

X_full = pd.concat([X_base, age_dummies], axis=1)

results = []
import statsmodels.api as sm
for prot in prot_cols:
    y = df[prot].astype(float)
    mask = (~y.isna()) & X_full.notnull().all(axis=1)
    if mask.sum() < 3:
        results.append((prot, np.nan, np.nan, np.nan, np.nan, mask.sum()))
        continue
    X = X_full.loc[mask, :]
    ysub = y.loc[mask]
    try:
        model = sm.OLS(ysub.values, X.values).fit()
        # coefficient for highest quartile vs Q1: find age_q_Q4 column
        age_col = [c for c in X.columns if 'age_q_Q4' in c]
        if len(age_col) == 0:
            # fallback to first age_q column
            age_col = [c for c in X.columns if c.startswith('age_q_')]
        if len(age_col) == 0:
            coef = se = t = pval = np.nan
        else:
            idx = list(X.columns).index(age_col[0])
            coef = model.params[idx]
            se = model.bse[idx]
            t = model.tvalues[idx]
            pval = model.pvalues[idx]
    except Exception as e:
        coef = se = t = pval = np.nan
    results.append((prot, coef, se, t, pval, mask.sum()))

res_df = pd.DataFrame(results, columns=['protein','beta','se','t','pvalue','n_nonmiss'])
res_df['adj_p'] = multipletests(res_df['pvalue'].fillna(1), method='fdr_bh')[1]
res_df = res_df.sort_values('adj_p')
res_df.to_csv(os.path.join(OUT_DIR, 'perprotein_ols_results.csv'), index=False)

# select top proteins
# if a protein list file exists, use it (one protein per line)
if os.path.exists(PROTEIN_LIST_FILE):
    with open(PROTEIN_LIST_FILE, 'r', encoding='utf-8') as fh:
        listed = [ln.strip() for ln in fh if ln.strip()]
    sel = [p for p in listed if p in df.columns]
    if len(sel) == 0:
        raise SystemExit('Protein list provided but none found in data columns')
else:
    sel = res_df.dropna(subset=['adj_p']).head(TOP_N)['protein'].values
    sel = [p for p in sel if p in df.columns]

if len(sel) == 0:
    raise SystemExit('No selected proteins found for heatmap')

expr_top = df[sel].copy()
# z-score per protein
expr_top_z = expr_top.apply(lambda col: (col - np.nanmean(col)) / np.nanstd(col, ddof=0), axis=0)

# column colors
lut = {'Q1':'#1f77b4','Q2':'#2ca02c','Q3':'#9467bd','Q4':'#ff7f0e','old':'#ff7f0e','young':'#1f77b4'}
col_colors = df['age_q'].map(lut)

# clustermap (proteins rows)
# use cividis colormap (good for colorblindness, perceptually uniform)
cmap_name = 'cividis'
g = sns.clustermap(expr_top_z.T, cmap=cmap_name, row_cluster=True, col_cluster=True, col_colors=col_colors.values, figsize=(10,8), xticklabels=False)
plt.suptitle('Top {} age-differential proteins (OLS adj covariates)'.format(len(sel)), y=1.02)
plt.savefig(os.path.join(OUT_DIR, 'age_diff_heatmap_ols.png'), dpi=300, bbox_inches='tight')
plt.close()

print('Done. Results and heatmap saved to', OUT_DIR)
